### ISSUE TITLE
The Legend of Zelda: Feature parity with ZeldaRandomizer 

### ISSUE THREAD

**Issue by wqferr** [ArchipelagoMW/Archipelago#1886]
### What feature would you like to see?

ZeldaRandomizer is a closed-source randomizer for TLoZ. Of particular interest is the number of [supported features](https://sites.google.com/site/zeldarandomizer/features). For example, the following features are listed on their website:

- Change the number of hearts required to get the items in caves that are normally white/magical sword caves in the normal game. 4-6 hearts are required for the white sword item, 10-14 for the magical sword.
- Scramble the items that are found in dungeons.  This means you may find the Red Ring in Level 1, or the Magical Book (as I understand, this is currently supported by Archipelago randomizer).
- The heart normally found on the coast (requiring a ladder) will be shuffled in with the dungeon items. This means that 2 hearts may be found in a dungeon, and a key item may require the ladder and be found on the coast.
- The white sword will be shuffled in with dungeon items.
- Randomly shuffle where the drops appear within a given dungeon.  All drops will still be present in the dungeon, but shuffled.  All rooms that naturally had items in the original game will still have them, but may be different (a heart container instead of a key, for instance).
- Randomly shuffles the rooms and doors within a dungeon. Dungeons will have the same layout as before (the "Eagle" still has the same shape), but the rooms may be placed in different spots (the Triforce room may be the first room you discover). Doors will be shuffled such that you will need approximately the same resources to fully complete the dungeons (in level 1, there will be 6 key doors and 2 bombable walls, for example); if additional doors are added to insure connectivity, the door may be either a bomb wall or an open door; this is even true of 2nd quest dungeons, which may normally have no bomb walls.
- Rooms may shuffle between dungeons.  This means you may find the skull room in Level 2, or the spiral staircase in Level 4.  There will still be one heart container/triforce/map/compass in levels 1-8, and a map/compass in level 9.  Gannon and Zelda will still be in level 9.

Along others.

I understand these are very much not trivial to implement, and I along with 2 other contributors will soon begin work understanding the ROM and implementing a couple simpler features to test the waters. Ideally we'd be able to reach feature parity within a couple months, but I shall hold the optimism at bay for now.

**Comment by t3hf1gm3nt** [comment]
Hello! One of the maintainers for TLOZ in AP here. Trust me, I would love nothing more than to have full feature parity with fcoughlin's version of the randomizer. But like you said, it is currently closed source so kinda hard to just take what is already done and implement it wholesale. Figured I could address the list you've written up here:

- Changing hearts requirement for the sword caves would be nice. Currently unsure how to hack the asm to make that functional
- Dungeon items are already shuffled like you've stated
- Ocean Ladder Heart is already shuffled as well
- White Sword is also shuffled into the pool
- Shuffling the item drops per dungeon could be doable now. Would just need map out the item pool into the individual dungeons. Low priority since this was developed with multiworld in mind, and letting all the item drops be loose is more interesting in a multiworld context, at least in my eyes and I believe @Rosalie-A 's, the main dev of TLOZ AP, as well.
- Dungeon Room shuffle is another thing that has been on the wishlist to do, just haven't taken the time to dig into the asm to see how it works. Will mention that entrance shuffle is also on said wishlist (and higher priority at least in my eyes).

The reason why none of these said issues is like you said: nontrivial to implement. TLOZ AP started as a hobby project of Roaslie's and I stepped in with some bugfixing and optimization to get it into the state it is currently in. Rosalie has been focusing on other projects and real life stuff since, and I've mostly just been handling maintenance as I haven't had much free time or willpower to work on it more.

This isn't me trying to defend whatever is here, just explaining the situation as to why TLOZ AP isn't as fleshed out as fcoughlin's. I actually would love the help in trying to reach feature parity, and am willing to help out where I can with any info you and your fellow contributors would need. Feel free to reach out to me here or on the AP discord.

**Comment by wqferr** [comment]
I'm studying a disassembly with comments ([a link to the disassembly here](https://github.com/aldonunez/zelda1-disassembly)) and one of the first things I stumbled upon was exactly the heart caves code. It had comments and variable names so (as far as I can tell) randomizing the hearts for these could be as simple as changing 2 easily findable bytes in the ROM.

**Comment by t3hf1gm3nt** [comment]
Yeah I believe that's the same disassembly Rosalie was using for reference. They did most of the current asm work. I only know just enough to get myself into trouble lol.

**Comment by wqferr** [comment]
Hopefully we can be of use to the team. I've never done anything related to romhacking before, though I do have plenty experience coding.

**Comment by wqferr** [comment]
I believe working with the FCEUX emulator would allow us to more quickly learn the ROM, due to its debug mode. BizHawk has memory watch, but no way to set breakpoints.

**Comment by Rosalie-A** [comment]
You might be interesting in browsing through [this](https://github.com/Rosalie-A/zelda1-disassembly), my fork of Aldo's disassembly that includes my changes made for the randomizer base patch. The base patch included in the TLoZ world is in fact made through assembling my version of the disasm and generating a patch against the base rom that Aldo's assembles into.

In general, most of the changes you've noted as well as the other differences with Fred's rando aren't of much interest to me, which is why they aren't implemented and won't be by me anytime soon. That said, I'm happy to answer any questions you may have and I'll gladly review a PR against this or my fork's repo for any changes you may wish to make. My only requirement is that changes be implemented as options where appropriate so people migrating between versions can receive the same experience as before with a minimum of fuss.

In general, the tools I used were Datacrystal's ROM and RAM map and the aforementioned disassembly. I didn't use a debugger since the disassembly is that well documented, but you may find it serves you well. The primary requested feature that I don't have is dungeon room shuffle and I can give you some basics that you may already know, but if not should jumpstart any progress you wish to make.

The Underworld in Z1 is composed of multiple 16x8 arrays. There are two sets: one for levels 1-6, and one for levels 7, 8, and 9. Each array contains one byte of data for each room, corresponding to its  geographical position in the array. In other words, each level is laid out in the array exactly as experienced in the ingame map. This is what leads to the [well known fact](https://ian-albert.com/games/legend_of_zelda_maps/zelda-dungeons.png) of the dungeons fitting together like puzzle pieces. An important fact to mention is that the "blank" spaces on the map are not unused: each one holds the data for either an item room or a transport room.

There's an array for room flags (dark, boss roar, major item if any), for enemies, doors and palette, for secrets, and of course for the rooms themselves, and I'm probably forgetting one too. You'll need to be conscious that you're changing all appropriate values. If you wanted an easy place to start, the actual room bytes is a good place. Each room byte is an index into the list of 42 defined room shapes, so you can play around with changing those and _generally_ things will work out. Notably, doors are not part of the room structure: those are handled by a different array. You can shuffle arbitrary rooms with arbitrary rooms and not run into too many issues aside from bonus Stepladder requirements and the two "corridor" rooms. 

(A tip so you don't fall into the same pitfall I did: Z1 loads the relevant underworld blocks into RAM. Start from a fresh file or make your save states outside the dungeon entrance if you want to see your changes live).

A final note is how Second Quest works, since you alluded to it in your post. Second Quest is two separate things. First, it's a series of patches onto the overworld, changing where entrances are, what they point to, and the occasional tile difference. Second of all, it's a completely different set of Underworld arrays. There is _zero_ overlap between First Quest Underworld and Second Quest Underworld in any sense, so changes made to one will not propagate to the other. This is why Second Quest just flat out isn't supported in my rando, because it would literally be doubling the work of locating addresses and designing logic. Not saying you can't go for it, just explaining why only First Quest is a thing in the rando.

If you have any questions about anything, feel free to ask here or on the Discord.

**Comment by wqferr** [comment]
Thank you! I will reach out once I have my foot in the mud and actually doing stuff, right now I'm just getting more familiar with the environment and the code.

**Comment by wqferr** [comment]
For future reference, work will be done in the following repository, on the branch Z1AP:

https://github.com/wqferr/zelda1-disassembly

**Comment by wqferr** [comment]
An update: [Reloaded Retro](https://www.reloadedretro.com/games/zelda)'s dev said they were interested in open sourcing their code, maybe as soon as next update.

They have most of the features of ZeldaRandomizer, plus entirely new ones. This could be huge in getting a better grasp at the ROM.
